src/sys/petsclib.cxx

```diff
@@ -47,5 +78,40 @@ void PetscLib::cleanup() {
   count = 0; // ensure that finalise is not called again later
 }
 
+void PetscLib::setPetscOptions(Options& options, const std::string& prefix) {
+  // Pass all options in the section to PETSc
+  for (auto& i : options.getChildren()) {
+    if (not i.second.isValue()) {
+      throw BoutException("Found subsection %s in %s when reading PETSc options - only "
+          "values are allowed in the PETSc options, not subsections",
+          i.first.c_str(), options.str().c_str());
+    }
+
+    // Note, option names in the input file don't start with "-", but need to be passed
+    // to PETSc with "-" prepended
+    auto petsc_option_name = "-"+prefix+i.first;
+    const char* value =
+      // "true" is the value given to an option with no value, when read from BOUT.inp
+      lowercase(i.second) == "true" ? nullptr :
+      // workaround to allow passing "true" to the petsc option
+      lowercase(i.second) == "petsc_true" ? "true" :
+      // allow this for symmetry with "petsc_true"
+      lowercase(i.second) == "petsc_false" ? "false" :
+      // default case: pass the value of the option
+      i.second.as<std::string>().c_str();
```

Comment:
Is the nested-ternary a good idea here? It looks to me like the neatest way to write this, but it might look a bit cryptic. In https://stackoverflow.com/questions/18237432/how-to-rewrite-complicated-lines-of-c-code-nested-ternary-operator different people have opposite opinions on whether it's good or bad.

Reply:
Would a switch be clearer?

Reply:
I guess it removes the ability to have const on the declaration.

Reply:
I'm just going to have a quick check, but I'm reasonably certain that we don't need a workaround to pass `"true"`, and it should be happy with `nullptr`.

This isn't too terrible. It is possible to have `const` with a `switch` using an immediately invoked lambda:

```cpp
const char* value = [i&]() {
   ...
  }();
```

but it would be even better if we could avoid needing the workaround here.

Reply:
I tried putting a `switch` and found the argument to `switch` has to be an integer, so `switch` isn't an option here.

Reply:
@ZedThree thanks - I got tired of trying to find in the PETSc docs a statement of all the things it might try to do with an option, so was trying to cover all bases.

Reply:
> I tried putting a `switch` and found the argument to `switch` has to be an integer, so `switch` isn't an option here.

Ah yes, good point!

Reply:
Ok I've had a play about with various options and I'm happy that no workaround is needed, and L97 & 99 can go.

I've also looked at the source code: https://www.mcs.anl.gov/petsc/petsc-current/src/sys/objects/options.c.html#PetscOptionsGetBool
and it does interpret `NULL` as truthy.

Reply:
:+1: nice, thanks @ZedThree!